File: lib/karabinex/command.ex
defmodule Karabinex.Command do
  alias Karabinex.Chord

  defstruct [:kind, :arg, :chord, :repeat]

  @type kind ::
          :app
          | :quit
          | :kill
          | :sh
          | :remap
          | :raycast

  @type spec ::
          {kind(), String.t()}
          | {kind(), String.t(), [option()]}

  @type option ::
          {:if, any()}
          | {:repeat, :key | :keymap}

  @type t :: %__MODULE__{
          chord: Chord.t(),
          kind: kind(),
          arg: String.t(),
          repeat: :key | :keymap | nil
        }

  def new(chord, kind, arg, opts \\ []) do
    %__MODULE__{
      chord: chord,
      kind: kind,
      arg: arg
    }
    |> add_opts(opts)
  end

  def add_opts(%__MODULE__{} = command, [{:repeat, :key} | rest]) do
    %{command | repeat: :key}
    |> add_opts(rest)
  end

  def add_opts(%__MODULE__{} = command, [{:repeat, :keymap} | rest]) do
    %{command | repeat: :keymap}
    |> add_opts(rest)
  end

  def add_opts(%__MODULE__{} = command, []), do: command
end

File: src/command.ts
import type { Chord } from "./chord";

export const commandKinds = ["app", "quit", "kill", "sh", "raycast"] as const;
export type CommandKind = (typeof commandKinds)[number];

export const repeatValues = ["key", "keymap"] as const;
export type RepeatValue = (typeof repeatValues)[number];

const commandKindSet = new Set<CommandKind>(commandKinds);
const repeatValueSet = new Set<RepeatValue>(repeatValues);

export function isCommandKind(value: string): value is CommandKind {
  return commandKindSet.has(value as CommandKind);
}

export function isRepeatValue(value: unknown): value is RepeatValue {
  return typeof value === "string" && repeatValueSet.has(value as RepeatValue);
}

export type CommandDef = {
  kind: CommandKind;
  arg: string;
  repeat?: RepeatValue;
};

export class Command {
  readonly chord: Chord;
  readonly kind: CommandKind;
  readonly arg: string;
  readonly repeat: boolean;

  constructor(chord: Chord, kind: CommandKind, arg: string, repeat = false) {
    this.chord = chord;
    this.kind = kind;
    this.arg = arg;
    this.repeat = repeat;
  }
}
